### PR TITLE
I understand how lazy queues behave on node startup

### PR DESCRIPTION
**Pull request by gerhard**
We've had quiet a few discussions around this, we're making our shared understanding public.

### CODE REVIEW

**Comment by gerhard** [comment]
I appreciate that the xml -> md makes the diff more difficult to understand, maybe side-by-side compare?

screen shot 2018-01-10 at 18 36 47

**Comment by gerhard** [comment]
We should probably close rabbitmq/rabbitmq-server#1284 when this gets accepted.

cc @carlhoerberg @Gsantomaggio

File: site/lazy-queues.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+and only load them in RAM when requested by consumers,
+therefore the lazy denomination.
+
+One of the main goals of lazy queues is to be able to support very long queues (many millions of messages).
+Queues can become very long for various reasons:
+
+<ul class="plain">
+  <li>consumers are offline / have crashed / are down for maintenance</li>
+  <li>there is a sudden message ingress spike, producers are outpacing consumers</li>
+  <li>consumers are simply slower than normal</li>
```
Comment (lukebakken) [inline site/lazy-queues.md]: Does "simply" add any meaning?

File: site/lazy-queues.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+</ul>
+
+By default, queues keep an in-memory cache of messages that's filled up as messages are published into RabbitMQ.
+The idea of this cache is to be able to deliver messages to consumers as fast as possible -
+note that persistent messages are written to disk as they enter the broker **and** kept in this cache at the same time.
```
Comment (lukebakken) [inline site/lazy-queues.md]: I suggest ending the previous sentence "as fast as possible." then start the new one "Note that..."

File: site/lazy-queues.md [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+Whenever the broker considers it needs to free up memory, messages from this cache will be paged out to disk.
+Paging messages to disk takes time and block the queue process, making it unable to receive new messages while it's paging.
```
Comment (lukebakken) [inline site/lazy-queues.md]: "takes time and blocks the queue process"

File: site/lazy-queues.md [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+Even if on recent RabbitMQ versions we have improved the paging algorithm,
```
Comment (lukebakken) [inline site/lazy-queues.md]: Suggestion: "Even though recent RabbitMQ versions include an improved paging algorithm, ..."

File: site/lazy-queues.md [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+One of the main goals of lazy queues is to be able to support very long queues (many millions of messages).
+Queues can become very long for various reasons:
+
+<ul class="plain">
+  <li>consumers are offline / have crashed / are down for maintenance</li>
+  <li>there is a sudden message ingress spike, producers are outpacing consumers</li>
+  <li>consumers are simply slower than normal</li>
+</ul>
+
+By default, queues keep an in-memory cache of messages that's filled up as messages are published into RabbitMQ.
```
Comment (lukebakken) [inline site/lazy-queues.md]: that's -> that is

File: site/lazy-queues.md [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+
+<pre class="sourcecode java">
+  Map&lt;String, Object> args = new HashMap&lt;String, Object>();
+  args.put("x-queue-mode", "lazy");
+  channel.queueDeclare("myqueue", false, false, false, args);
+</pre>
+
+### Configuration using policy
+
+To specify a queue mode using a policy, add the key `queue-length` to a policy definition, e.g.:
```
Comment (lukebakken) [inline site/lazy-queues.md]: `queue-length` -> `queue-mode`

File: site/lazy-queues.md [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+make run ARGS="-y 0 -s 1000 -f persistent -C 1000000 -u lazy -qa x-queue-mode=lazy -ad false"
+</pre>
+
+**Note that this was a very simplistic test.**
+**Please make sure to run your own benchmarks.**
+
+### Converting between queue modes
+
+If we need to convert a `default` queue into a `lazy` one,
+then we will suffer the same performance impact as when a queue needs to page messages to disk.
```
Comment (lukebakken) [inline site/lazy-queues.md]: When converting a `default` queue into a `lazy` one, the operation will suffer the same performance impact as when a queue needs to page messages to disk.

File: site/lazy-queues.md [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+
+When we convert a `default` queue into a `lazy` one,
+the queue will first page messages to disk,
+and then it will start accepting publishes, acks, and other commands.
```
Comment (lukebakken) [inline site/lazy-queues.md]: During conversion from a `default` queue into a `lazy` one, the queue...

File: site/lazy-queues.md [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+These messages will use **63MB** of system memory.
+The queue process will use another **8.4MB** of system memory, bringing the total to just over **70MB**.
+This is an important consideration for capacity planning if the RabbitMQ node is memory constrained,
+or if there are many lazy queues running on this RabbitMQ node.
+
+**It is important to remember that an under-provisioned RabbitMQ node in terms of memory or disk space will fail to start**
+
+To prevent any messages in a lazy queue loading into memory on node startup,
+you can set the `queue_index_embed_msgs_below` to `0`, as described in [RabbitMQ Configuration](configure.html).
+We encourage you to read [Persistence Configuration](persistence-conf.html) before tuning this property.
```
Comment (lukebakken) [inline site/lazy-queues.md]: 🎉 👍

File: site/lazy-queues.md [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+  <tr>
+    <th>rabbitmqctl</th>
+    <td>
+      <pre>rabbitmqctl set_policy Lazy "^lazy-queue$" '{"queue-mode":"lazy"}' --apply-to queues</pre>
+    </td>
+  </tr>
+  <tr>
+    <th>rabbitmqctl (Windows)</th>
+    <td>
+      <pre>rabbitmqctl set_policy Lazy "^lazy-queue$" "{""queue-mode"":""lazy""}" --apply-to queues</pre>
```
Comment (lukebakken) [inline site/lazy-queues.md]: Windows shell quoting = 🤦‍♂️

Comment (gerhard) [reply site/lazy-queues.md]: 🚶

**Pull request merged**